src/test/kotlin/pt/up/fe/ni/website/backend/controller/RoleControllerTest.kt

```diff
@@ -0,0 +1,720 @@
+            )
+        }
+    }
+
+    @NestedTest
+    @DisplayName("GET /roles/{id}")
+    inner class GetSpecificRole {
+        @BeforeEach
+        fun addRole() {
+            roleRepository.save(testRole)
+        }
+
+        @Test
+        fun `should return testRole when provided by its id`() {
+            mockMvc.perform(get("/roles/${testRole.id}")).andExpectAll(
+                status().isOk,
+                content().contentType(MediaType.APPLICATION_JSON),
+                content().json(objectMapper.writeValueAsString(testRole))
+            ).andDocument(
+                documentationRoles,
+                "Returns a specific role",
+                "Returns a summary brief of a specific role, which makes getting one role way more efficient. "
+            )
+        }
+
+        @Test
+        fun `should return error on invalid roleID`() {
+            mockMvc.perform(get("/roles/4020")).andExpectAll(
+                status().isNotFound(),
+                content().contentType(MediaType.APPLICATION_JSON),
+                jsonPath("$.errors.length()").value(1),
+                jsonPath("$.errors.[0].message").value("role not found with id 4020")
+            ).andDocumentErrorResponse(documentationRoles, hasRequestPayload = true)
+        }
+    }
+
+    @NestedTest
+    inner class CreateNewRole {
+
+        @BeforeEach
+        fun addRole() {
+            generationRepository.save(testGeneration)
+            roleRepository.save(testRole)
+            testGeneration.roles.add(testRole)
+            roleRepository.save(testRole)
+            generationRepository.save(testGeneration)
+        }
+
+        @AfterEach
+        fun removeRole() {
+            testGeneration.roles.remove(testRole)
+            testGeneration.roles.remove(otherTestRole)
+            roleRepository.save(testRole)
+            generationRepository.save(testGeneration)
+        }
+
+        @Test
+        fun `should add new role`() {
+            mockMvc.perform(
+                post("/roles/new")
+                    .contentType(MediaType.APPLICATION_JSON)
+                    .content(
+                        objectMapper.writeValueAsString(
+                            mapOf(
+                                "name" to otherTestRole.name,
+                                "permissions" to otherTestRole.permissions.map { it.bit }.toList(),
+                                "isSection" to otherTestRole.isSection,
+                                "associatedActivities" to otherTestRole.associatedActivities,
+                                "accounts" to otherTestRole.accounts
+                            )
+                        )
+                    )
+            ).andExpectAll(
+                status().isOk,
+                content().contentType(MediaType.APPLICATION_JSON),
+                jsonPath("$.name").value(otherTestRole.name),
+                jsonPath("$.id").value(2), // this must be hardcoded
```

Comment:
you could just verify that it exists to prevent future mistakes. It doesn't really matter and the id rule could be something other than sequential